Improve GOC score calculation

ipilizota opened the pull request:
## Description

Improving the current GOC score calculation

**Related JIRA tickets:**
- ENSCOMPARASW-5668

## Overview of changes

#### Change 1
- For a gene in `neighbourhood1`, instead of searching for orthologous matches in `neighbourhood2` starting from the `previous_match+1` position onwards, search throughout the whole `neighbourhood2`.

#### Change 2
- Ignore the position of matches. Previously: A match was considered only if it was within certain distance from a previous match (previous match was a match for one of the previously searched genes from `neighbourhood1`). Now: Any match is considered, given the directionality criteria is satisfied.

## Testing
`pytest`

ipilizota commented on scripts/pipeline/orthology_benchmark.py:
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
+                shared_orthologs += 1
                 break  # Don't look for multiple matches
 
     return (shared_orthologs / (2 * n)) * 100
```
We were talking about keeping it simple and dividing by `2*n` which is the maximum number of neighbours fetched for `gene1`. But `gene1` can have fewer neighbours due to the fragmented assembly or simply by being the first or last one on the chromosome. Based on my code it is easy to divide by the number of neighbours we actually have for `gene1` but that has its pros and cons.

ipilizota replied:
@dthybert Not sure if "👍" means to keep it as it is or divide by the actual number of neighbours?

thiagogenez commented on scripts/pipeline/orthology_benchmark.py:
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
     strand2 = neighbourhood2.iloc[index2]["strand"]
```
There is straight access to the data frame array before checking the existence of indexes in the array. Would a safety sanity check need here?e

ipilizota replied:
You're right. I should add safety sanity checks (maybe not just at this place) and then test cases accordingly. That will also address your comment in the test script 🙂

thiagogenez commented on src/python/tests/test_orthology_benchmark.py:
```diff
@@ -515,44 +515,37 @@ def test_get_neighbours(gene: str, n: int, exp_genes: List[Tuple[str, str, float
+         2, 100)
     ]
 )
 def test_calculate_goc_genes(gene1: str, gene2: str, neighbours1: List[Tuple[str, str, int, str]],
```
Does the test only assume valid inputs (with valid data and invalid data)?  If affirmative, should be interesting to include invalid inputs as well (i.e. empty data frame or dataframe with negative values)?

ipilizota replied:
Well spotted 🙂 As I said above in one of my comments, I will address this. Thank you!

thiagogenez reviewed (commented):
Make a few comments regarding the code

dthybert commented on scripts/pipeline/orthology_benchmark.py:
```diff
@@ -544,47 +542,41 @@ def calculate_goc_genes(gene1: str, gene2: str, neighbourhood1: pandas.DataFrame
         neighbourhood2.drop("index", 1, inplace=True)
+        index2 = neighbourhood2.index[neighbourhood2["gene_id"] == gene2][0]
         strand_mismatch = True
     else:
         strand_mismatch = False
 
```
yes I agree it will be more readable and the code easier to extend if required in the future

ipilizota replied:
Will do 🙂 Thank you 🙂

dthybert reviewed (approved):
I approve with the assumption that you will take Thiago comment into account before merging

ipilizota commented:
@thiagogenez 

> try to run
> 
> ```
> git cherry-pick 95c1fc8c6124ca82e4ee33ba140b54f4b3d45039
> ```
> 
> to solve the `pylint` problem

```
fatal: bad object 95c1fc8c6124ca82e4ee33ba140b54f4b3d45039
```
I'll get back to it tomorrow and find a way. Thank you anyways 🙂

thiagogenez commented:
Travis `Pylint` job is failing on all branches. I need to check it out.

thiagogenez commented:
Regarding Travis `Perl external unit tests on the latest version, with code coverage` you need to either rebase your branch with `main` or edit the file `travisci/perl-external_unittest_harness.sh` similar to @CristiGuijarro 's [commit](https://github.com/Ensembl/ensembl-compara/commit/74402b87cdc2b25b95d42c611680aa2849aa3094)

ipilizota commented:
Thank you @thiagogenez. Yeah, I have to rebase it against `main` and I expect that will sort out the Perl side.
`pylint` 😭

thiagogenez reviewed (approved):
Please just read my comments in the main thread conversation.

ipilizota commented:
> Please just read my comments in the main thread conversation.

You mean that I should ignore your suggestions in the code? 🙊 I have already implemented safety sanity checks in the script (haven't pushed them yet) and I was planning to add some test cases for them. I see you've already approved this PR. Do you mind if I ping you to have a look at this PR again once I push everything regarding safety sanity checks?

ipilizota commented:
@thiagogenez 🙊
So... I included safety sanity checks and then `pylint` complained: `R0912: Too many branches (13/12) (too-many-branches)`. So I refactored the code, i.e. added one more function and corresponding test. The only suggestion that you made and I didn't incorporate was to use negative start positions for genes. That info is not used in this function at all.

I also had to edit a bit some old tests because `pylint` started complaining about them.

thiagogenez reviewed (approved):
Recommended improvement suggestions, such as extra sanity checks and tests, have been nicely addressed and implemented.

ipilizota commented:
Travis is failing on `pylint` - being addressed in https://github.com/Ensembl/ensembl-compara/pull/482 & will be pulled on the project branch upon PR approval.
